pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,417 @@
+	klog.V(2).InfoS("Processing Namespace UPDATE event", "Namespace", klog.KObj(ns))
+
+	affectedPods := l7c.getPodsFromNS(ns)
+	for _, pod := range affectedPods {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		l7c.queue.Add(podNN)
+	}
+}
+
+func (l7c *l7Controller) deleteNamespace(old interface{}) {
+	namespace, ok := old.(*v1.Namespace)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Error decoding object when deleting Namespace, invalid type", "object", old)
+			return
+		}
+		namespace, ok = tombstone.Obj.(*v1.Namespace)
+		if !ok {
+			klog.ErrorS(nil, "Error decoding object tombstone when deleting Namespace", "invalid obj", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).InfoS("Processing Namespace %s DELETE event, labels: %v", namespace.Name, namespace.Labels)
+
+	affectedPods := l7c.getPodsFromNS(namespace)
+	for _, pod := range affectedPods {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		l7c.queue.Add(podNN)
+	}
+}
+
+func (l7c *l7Controller) getPodsFromNS(ns *v1.Namespace) []*v1.Pod {
+	var (
+		selectedPods     []*v1.Pod
+		nonAnnotatedPods []*v1.Pod
+	)
+	podSelector := labels.Everything()
+	pods, err := l7c.podLister.Pods(ns.Name).List(podSelector)
+	if err != nil {
+		klog.ErrorS(err, "could not list Pods for given namespace", "namespace", ns)
+		return nil
+	}
+	selectedPods = append(selectedPods, pods...)
+
+	// Only select the non annotated Pods, as annotated Pods are handled separately
+	for _, pod := range selectedPods {
+		_, ok := pod.Annotations[l7VisibilityAnnotationKey]
+		if !ok {
+			nonAnnotatedPods = append(nonAnnotatedPods, pod)
+		}
+	}
+	return nonAnnotatedPods
+}
+
+func (l7c *l7Controller) syncPodEvent(podNN string) error {
+	var (
+		pod             *v1.Pod
+		err             error
+		annotationValue string
+	)
+	podNamespace, podName := k8s.SplitNamespacedName(podNN)
+	if pod, err = l7c.podLister.Pods(podNamespace).Get(podName); err != nil {
+		// Remove the TC flows if the Pod has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+
+	// TCmarkflow will only be removed if both the Pod and Namespace annotations
+	// do not contain the L7 annotation
+	annotationValue, ok := pod.Annotations[l7VisibilityAnnotationKey]
+	if !ok {
+		podNamespace, err := l7c.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			return l7c.removeTCFlow(podNN)
+		} else {
+			annotationValue, ok = podNamespace.Annotations[l7VisibilityAnnotationKey]
+			if !ok {
+				return l7c.removeTCFlow(podNN)
+			}
+		}
+	}
+
+	// Check if the annotation value is one of the specified values
+	var direction v1alpha2.Direction
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return fmt.Errorf(errInvalidAnnotation)
+	}
+	tcName := fmt.Sprintf("tcl7-%s", podNN)
```

Comment:
If tcName is just transformed from pod name, there is no point to store the value in a map. It could just declare a set[string] for installed pods.

Reply:
Changed it to L7PodNNSet